### ISSUE TITLE
[Question] how to fix the inference outcome?

### ISSUE THREAD

**Issue by nevermet** [Lightning-AI/lit-gpt#690]
Dear all,

I finetuned the model with my own data and ran the generate/lora.py with the test set.

I found that the output is different every time I ran the code... can I make the outcome consistent? With such as seed numbers?

Thank you in advance.

**Comment by Andrei-Aksionov** [comment]
Hello @nevermet 

Try to use `fabric.seed_everything` function as it's done in fine-tune script:
https://github.com/Lightning-AI/lit-gpt/blob/f6841380169000e225f45be22187862409bae185/finetune/lora.py#L102

**Comment by Andrei-Aksionov** [comment]
Ok, maybe I've spent too much time on such an easy task, nevertheless wanted to double-check that everything loads for LoRA variant without issues.

@nevermet The answer is the same: you can use `seed_everything` method like it's done in `generate/base.py`:
https://github.com/Lightning-AI/lit-gpt/blob/f6841380169000e225f45be22187862409bae185/generate/base.py#L179

@carmocca I've noticed that out of all generation scripts, the seed is fixed only in `generate/base.py`. Is it by mistake or on purpose? Do we need to apply it to all scripts for consistency?

**Comment by nevermet** [comment]
Thank you for your answers. 

I think that generate/lora.py uses generate/base.py which has the code 'L.seed_everything()'. Then do I have do something in generate/lora.py? Or do I have to just change the temperature value to a very small number?

Thank you.

**Comment by Andrei-Aksionov** [comment]
> I think that generate/lora.py uses generate/base.py which has the code 'L.seed_everything()'. 

No, `L.seed_everything()` is in the `generate/base.py::main` function; `generate/lora.py` utilizes `generate/base.py::generate`.

------

> Or do I have to just change the temperature value to a very small number?

Or you can specify `top_k` equal to 1. In both cases the sampling will be done from a set with a single logit, that has the highest probability.
FIY: here is a good [answer](https://ai.stackexchange.com/questions/32477/what-is-the-temperature-in-the-gpt-models) if you are not familiar with the concept of `temperature`.
But there is a caveat: you need to provide `--precision 32-true`. Otherwise it will not work. Based on my experience it's caused by `kv cache`: with not full precision there is a small deviation after n-first tokens.

Overall my recommendation is to fix the seed. There is a reason why we want to sample from some distribution and to not output the logit with the highest probability (quality of the output might suffer).

-------

> Also, doesn't the number in seed_everything() matter? Why 1337 in finetune and why 1234 in generate?

Not sure why it is, but it shouldn't matter. You can put `42` if you wish in both cases 😉.

**Comment by carmocca** [comment]
> @carmocca I've noticed that out of all generation scripts, the seed is fixed only in generate/base.py. Is it by mistake or on purpose? Do we need to apply it to all scripts for consistency?

It's probably a mistake. But it doesn't matter. It could be removed or added from all generation scripts.
